src/acl/HttpStatus.cc

```diff
@@ -48,9 +48,9 @@ int acl_httpstatus_data::compare(acl_httpstatus_data* const& a, acl_httpstatus_d
     if (ret == 0) {
         const SBuf sa = a->toStr();
         const SBuf sb = b->toStr();
-        debugs(28, DBG_CRITICAL, "WARNING: '" << sa << "' is a subrange of '" << sb << "'");
-        debugs(28, DBG_CRITICAL, "WARNING: because of this '" << sa << "' is ignored to keep splay tree searching predictable");
-        debugs(28, DBG_CRITICAL, "WARNING: You should probably remove '" << sb << "' from the ACL named '" << AclMatchedName << "'");
+        debugs(28, DBG_CRITICAL, "WARNING: '" << sa << "' is a subrange of '" << sb << "'" <<
+                Debug::Extra << "Because of this '" << sa << "' is ignored to keep splay tree searching predictable" <<
+                Debug::Extra << "You should probably remove '" << sb << "' from the ACL");
     }
```

Comment:
I do not understand how the official code comes up with that "a is a subrange of b" conclusion. AFAICT, there are four possible outcomes inside this `if` statement:

1. a is b: OK, even though some admins will find the term "subrange" misleading in this case.
2. a is a subrange of b: OK, we are ignoring a smaller, duplicated a and are telling the truth about it.
3. b is a subrange of a: Bad, we are going to ignore a unique sub-range of a and lie about doing something else.
4. a partially overlaps with b: Bad, we are going to ignore a unique sub-range of a and lie about doing something else.

Here is my math:

```

cmp(b,a) returns zero:
        b1 >= a1 && b1 <= a2 
        (partial overlap or b is included in a)

cmp(b,a) returns +1 but cmp(a,b) returns 0:
        b1 < a1 && (a1 >= b1 && a1 <= b2)  =>
        b1 < a1 && b2 >= a1
        (partial overlap or a is included in b)

cmp(b,a) returns -1 but cmp(a,b) returns 0:
        b1 >= a1 && b1 > a2 && (a1 >= b1 && a1 <= b2)  =>
        b1 == a1 && b1 > a2 && a1 <= b2  => (dropping b1==a1 but replacing b1 with a1 and a1 with b1) =>
        a1 > a2 && b1 <= b2  =>
        a1 > a2  =>
        impossible!

... where cmp(x,y) is aclHTTPStatusCompare(x,y)
```

Are my math and conclusions correct or did I make a mistake somewhere?

Reply:
I edited the previous comment to add problematic master-based output that appears to confirm my math/suspicions. Just to clarify: I am not requesting any code changes at this point; just a confirmation that my logic is sound (and Squid is very buggy in this area).

Reply:
ACLIP ACLs (`src`, `dst`, and `localip`) are similarly broken:

```
Processing: acl test1 src 127.0.0.1  127.0.0.0-127.0.0.255
WARNING: (B) '127.0.0.1' is a subnetwork of (A) '127.0.0.0-127.0.0.255'
WARNING: because of this '127.0.0.0-127.0.0.255' is ignored to keep splay tree searching predictable
WARNING: You should probably remove '127.0.0.1' from the ACL named 'test1'
```

```
Processing: acl test3 src 127.0.0.0-127.0.0.128  127.0.0.128-127.0.0.255
WARNING: (A) '127.0.0.128-127.0.0.255' is a subnetwork of (B) '127.0.0.0-127.0.0.128'
WARNING: because of this '127.0.0.128-127.0.0.255' is ignored to keep splay tree searching predictable
WARNING: You should probably remove '127.0.0.128-127.0.0.255' from the ACL named 'test3'
```

I am working on a fix.